src/transaction/log_replication.cpp

```diff
@@ -1,173 +1,264 @@
+  {
+    assert (m_redo_lsa < end_redo_lsa);
+
+    // redo all records from current position (m_redo_lsa) until end_redo_lsa
+
+    // make sure the log page is refreshed. otherwise it may be outdated and new records may be missed
+    m_reader.set_lsa_and_fetch_page (m_redo_lsa, log_reader::fetch_mode::FORCE);
+
+    while (m_redo_lsa < end_redo_lsa)
+      {
+	// read and redo a record
+	m_reader.set_lsa_and_fetch_page (m_redo_lsa);
+
+	const log_rec_header header = m_reader.reinterpret_copy_and_add_align<log_rec_header> ();
+
+	switch (header.type)
+	  {
+	  case LOG_REDO_DATA:
+	    read_and_redo_record<log_rec_redo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_MVCC_REDO_DATA:
+	    read_and_redo_record<log_rec_mvcc_redo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_UNDOREDO_DATA:
+	  case LOG_DIFF_UNDOREDO_DATA:
+	    read_and_redo_record<log_rec_undoredo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_MVCC_UNDOREDO_DATA:
+	  case LOG_MVCC_DIFF_UNDOREDO_DATA:
+	    read_and_redo_record<log_rec_mvcc_undoredo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_RUN_POSTPONE:
+	    read_and_redo_record<log_rec_run_postpone> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_COMPENSATE:
+	    read_and_redo_record<log_rec_compensate> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_DBEXTERN_REDO_DATA:
+	  {
+	    log_rec_dbout_redo dbout_redo = m_reader.reinterpret_copy_and_add_align<log_rec_dbout_redo> ();
+	    log_rcv rcv;
+	    rcv.length = dbout_redo.length;
+	    log_rv_redo_record (&thread_entry, m_reader, RV_fun[dbout_redo.rcvindex].redofun, &rcv, &m_redo_lsa, 0,
+				nullptr, m_redo_unzip);
+	  }
+	  break;
+	  default:
+	    // do nothing
+	    break;
+	  }
+
+	{
+	  std::unique_lock<std::mutex> lock (m_redo_lsa_mutex);
+	  m_redo_lsa = header.forw_lsa;
+	}
+
+        // TODO: why is this notification inside this loop? where this function is
+        // called, can we be sure that 'm_redo_lsa == nxio_lsa' ?
+
+	// notify who waits for end of replication
+	if (m_redo_lsa == end_redo_lsa)
+	  {
+	    m_redo_condvar.notify_all ();
+	  }
+      }
+  }
+
+  template <typename T>
+  void
+  replicator::read_and_redo_record (cubthread::entry &thread_entry, LOG_RECTYPE rectype, const log_lsa &rec_lsa)
+  {
+    m_reader.advance_when_does_not_fit (sizeof (T));
+    const T log_rec = m_reader.reinterpret_copy_and_add_align<T> ();
+
+    // To allow reads on the page server, make sure that all changes are visible.
+    // Having log_Gl.hdr.mvcc_next_id higher than all MVCCID's in the database is a requirement.
+    MVCCID mvccid = log_rv_get_log_rec_mvccid (log_rec);
+    if (mvccid != MVCCID_NULL && !MVCC_ID_PRECEDES (mvccid, log_Gl.hdr.mvcc_next_id))
+      {
+	log_Gl.hdr.mvcc_next_id = mvccid;
+	MVCCID_FORWARD (log_Gl.hdr.mvcc_next_id);
+      }
+
+    const VPID rec_vpid = log_rv_get_log_rec_vpid<T> (log_rec);

```

Comment:
Isn't the next part identical to log_rv_redo_record_sync_or_dispatch_async?

Reply:
It is. Except for the `#ifdef (SERVER_MODE)` bit